### PR TITLE
CatalogKey can be used to index catalog

### PR DESCRIPTION
**Pull request by AlpAribal**
# Description

Makes it possible to get, set, and delete sources in a catalog using a `CatalogKey` or a `(id, version)` tuple.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring including code style reformatting 
- [ ] Other (please describe):

# Checklist:

- [ ] I have read the [contributing guideline doc](https://squirrel-core.readthedocs.io/en/latest/usage/code_of_conduct.html) (external contributors only)
- [x] Lint and unit tests pass locally with my changes
- [x] I have kept the PR small so that it can be easily reviewed  
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All dependency changes have been reflected in the pip requirement files.

### CODE REVIEW

File: squirrel/catalog/catalog.py [hunk 1]
```diff
@@ -71,22 +72,54 @@ def __eq__(self, other: Any) -> bool:  # noqa D105
...
         if identifier not in self:
-            # return a dummy object to let the user set a version directly
-            return DummyCatalogSource(identifier, self)
-        return self.sources[identifier][-1]
+            if isinstance(identifier, str):
+                # return a dummy object to let the user set a version directly
+                return DummyCatalogSource(identifier, self)
+            else:
+                # DummyCatalogSource only allows setting the version after initialization, cannot set it at this point
+                raise KeyError
```
Comment (AlirezaSohofi) [inline squirrel/catalog/catalog.py:118]: please give an informative error message

Comment (AlirezaSohofi) [reply squirrel/catalog/catalog.py:118]: would be cool to have the test case for this too

File: squirrel/catalog/catalog.py [hunk 2]
```diff
@@ -71,22 +72,54 @@ def __eq__(self, other: Any) -> bool:  # noqa D105
...
+            version = -1
+        else:
+            identifier, version = identifier
 
-    def __getitem__(self, identifier: str) -> CatalogSource:  # noqa D105
         if identifier not in self:
-            # return a dummy object to let the user set a version directly
-            return DummyCatalogSource(identifier, self)
-        return self.sources[identifier][-1]
+            if isinstance(identifier, str):
```
Comment (AlirezaSohofi) [inline squirrel/catalog/catalog.py:113]: why do we need this additional check `if isinstance(identifier, str):` ?

File: squirrel/catalog/catalog.py [hunk 3]
```diff
@@ -71,22 +72,54 @@ def __eq__(self, other: Any) -> bool:  # noqa D105
...
-    def __delitem__(self, identifier: Union[str, CatalogKey]) -> None:  # noqa D105
-        del self._sources[identifier]
+    def __contains__(self, identifier: Union[str, CatalogKey, Tuple[str, int]]) -> bool:  # noqa D105
+        if isinstance(identifier, str):
+            return identifier in self._sources
+
+        identifier, version = identifier
+        return identifier in self._sources and version in self._sources[identifier]
+
+    def __delitem__(self, identifier: Union[str, CatalogKey, Tuple[str, int]]) -> None:  # noqa D105
```
Comment (AlirezaSohofi) [inline squirrel/catalog/catalog.py:82]: the behaviour might be unexpected, you can overwrite an existing version, or set and arbitrary version, etc, should we explain it some where?

File: test/test_catalog/test_catalog.py [hunk 4]
```diff
@@ -22,16 +23,25 @@ def test_catalog_createsimple() -> None:
 
 
 def test_catalog_versioning() -> None:
-    """Test versioning source in a Catalog"""
+    """Test versioning a source in a Catalog using all allowed identifier types."""
```
Comment (AlirezaSohofi) [inline test/test_catalog/test_catalog.py:26]: please add test cases for `__delitem__`, `__contains__`, `__getitem__` for non-existing CatalogKey, and providing unsupported CatalogKey (e.g. non-string identifier?)

**Review by AlirezaSohofi — CHANGES_REQUESTED** [review]

File: squirrel/catalog/catalog.py (hunk 2 above)
Comment (AlpAribal) [reply squirrel/catalog/catalog.py:113]: We allow two types of identifiers:
1) `str`: Just the id of the catalog source, the user gets the version they want later on, e.g. `cat["my_id"][my_ver]`
2) `Tuple[str, int]`: This is the new format, basically either a `CatalogKey` or a tuple similar to it. This is what this PR introduces. The user can specify the version directly, e.g. `cat["my_id", my_ver]`.

To differentiate btw the two types, checking if identifier is a string is easier. If we check `isinstance(identifier, CatalogKey)`, then we do not cover the `Tuple[...]` case, and if we check `isinstance(identifier, Tuple)` then we do not cover the `CatalogKey` case. So, we check if identifier is string and if not we conclude that the identifier is tuple-like.

File: squirrel/catalog/catalog.py (hunk 3 above)
Comment (AlpAribal) [reply squirrel/catalog/catalog.py:82]: Which behavior exactly? The string identifier case is same as before. Do you mean we should remove all version also when a tuple identifier is provided?

**Review by AlirezaSohofi — DISMISSED** [review]
LG :+1:

**Pull request merged**